File: examples/anthropic-managed-agents/python/anthropic_managed_agents_e2b/sandbox_worker.py
from __future__ import annotations

import logging
import shlex
from pathlib import Path

from e2b import Sandbox

from anthropic_managed_agents_e2b.app_sandbox_store import JsonSandboxStore
from anthropic_managed_agents_e2b.environment import (
    WEBHOOK_SANDBOX_METADATA_KEY,
    WEBHOOK_SANDBOX_STORE_METADATA_KEY,
    WORKER_SANDBOX_METADATA_KEY,
    WORKER_SANDBOX_STORE_METADATA_KEY,
    add_sandbox_to_metadata_store,
    clear_matching_sandbox_metadata,
)
from anthropic_managed_agents_e2b.settings import PACKAGE_ROOT, Settings

REMOTE_DIR = "/opt/anthropic-managed-agents"
REMOTE_WORKDIR = "/mnt/session"
REMOTE_PACKAGE_DIR = f"{REMOTE_DIR}/anthropic_managed_agents_e2b"
REMOTE_WORKER = f"{REMOTE_DIR}/worker.py"
REMOTE_PID = f"{REMOTE_DIR}/worker.pid"
REMOTE_LOG = f"{REMOTE_DIR}/worker.log"
REMOTE_WEBHOOK_PID = f"{REMOTE_DIR}/webhook.pid"
REMOTE_WEBHOOK_LOG = f"{REMOTE_DIR}/webhook.log"
REMOTE_ENVIRONMENT_ID = f"{REMOTE_WORKDIR}/.anthropic-environment-id"
REMOTE_ENVIRONMENT_KEY = f"{REMOTE_WORKDIR}/.anthropic-environment-key"
REMOTE_WEBHOOK_SIGNING_KEY = f"{REMOTE_WORKDIR}/.anthropic-webhook-signing-key"
REMOTE_WORKER_MAX_IDLE_SECONDS = f"{REMOTE_WORKDIR}/.worker-max-idle-seconds"
REMOTE_LOG_LEVEL = f"{REMOTE_WORKDIR}/.log-level"
logger = logging.getLogger(__name__)
REMOTE_WORKER_ENTRYPOINT = """\
from anthropic_managed_agents_e2b.worker_runtime import main

if __name__ == "__main__":
    main()
"""


def create_or_connect_worker_sandbox(
    settings: Settings,
    *,
    template_name: str,
    timeout_seconds: int,
    sandbox_id: str | None = None,
) -> Sandbox:
    if sandbox_id:
        return Sandbox.connect(sandbox_id, timeout=timeout_seconds)

    return Sandbox.create(
        template_name,
        timeout=timeout_seconds,
        lifecycle={"on_timeout": "pause", "auto_resume": True},
        metadata={
            "managed_by": "anthropic-managed-agents-e2b",
            "anthropic.environment_id": settings.anthropic_environment_id or "",
        },
    )


def upload_worker(sandbox: Sandbox) -> None:
    sandbox.commands.run(f"mkdir -p {shlex.quote(REMOTE_PACKAGE_DIR)}", timeout=15)
    sandbox.files.write(REMOTE_WORKER, REMOTE_WORKER_ENTRYPOINT)
    sandbox.files.write(
        f"{REMOTE_PACKAGE_DIR}/__init__.py",
        (PACKAGE_ROOT / "__init__.py").read_text(),
    )
    sandbox.files.write(
        f"{REMOTE_PACKAGE_DIR}/worker_runtime.py",
        (PACKAGE_ROOT / "worker_runtime.py").read_text(),
    )
    sandbox.files.write(
        f"{REMOTE_PACKAGE_DIR}/webhook_runtime.py",
        (PACKAGE_ROOT / "webhook_runtime.py").read_text(),
    )


def start_worker_process(
    sandbox: Sandbox,
    settings: Settings,
    *,
    worker_max_idle_seconds: float | None,
    log_level: str,
) -> None:
    envs = {
        "ANTHROPIC_ENVIRONMENT_ID": settings.require_anthropic_environment_id(),
        "ANTHROPIC_ENVIRONMENT_KEY": settings.require_anthropic_environment_key(),
        "WORKER_MAX_IDLE_SECONDS": "none"
        if worker_max_idle_seconds is None
        else str(worker_max_idle_seconds),
        "LOG_LEVEL": log_level,
    }
    wrapper = f"""
    set -eu
    cd /mnt/session
    nohup python {shlex.quote(REMOTE_WORKER)} > {shlex.quote(REMOTE_LOG)} 2>&1 < /dev/null &
    printf '%s\\n' "$!" > {shlex.quote(REMOTE_PID)}
    """
    result = sandbox.commands.run(f"bash -lc {shlex.quote(wrapper)}", envs=envs, timeout=15)
    if result.exit_code not in (0, None):
        raise RuntimeError(f"worker start failed:\n{result.stdout}\n{result.stderr}")


def worker_process_is_running(sandbox: Sandbox) -> bool:
    check = (
        f"test -f {shlex.quote(REMOTE_PID)} && "
        f'pid="$(cat {shlex.quote(REMOTE_PID)})" && '
        'test -n "$pid" && '
        'kill -0 "$pid"'
    )
    try:
        result = sandbox.commands.run(f"bash -lc {shlex.quote(check)}", timeout=5)
    except Exception:
        return False
    return result.exit_code == 0


def ensure_worker_process(
    sandbox: Sandbox,
    settings: Settings,
    *,
    worker_max_idle_seconds: float | None,
    log_level: str,
) -> None:
    upload_worker(sandbox)
    if worker_process_is_running(sandbox):
        return
    start_worker_process(
        sandbox,
        settings,
        worker_max_idle_seconds=worker_max_idle_seconds,
        log_level=log_level,
    )


def start_worker_sandbox(
    settings: Settings,
    *,
    template_name: str,
    timeout_seconds: int,
    worker_max_idle_seconds: float | None,
    log_level: str,
    sandbox_id: str | None = None,
) -> Sandbox:
    settings.require_anthropic_environment_id()
    settings.require_anthropic_environment_key()
    sandbox = create_or_connect_worker_sandbox(
        settings,
        template_name=template_name,
        timeout_seconds=timeout_seconds,
        sandbox_id=sandbox_id,
    )
    ensure_worker_process(
        sandbox,
        settings,
        worker_max_idle_seconds=worker_max_idle_seconds,
        log_level=log_level,
    )
    if settings.anthropic_api_key:
        add_sandbox_to_metadata_store(
            api_key=settings.anthropic_api_key,
            environment_id=settings.require_anthropic_environment_id(),
            legacy_key=WORKER_SANDBOX_METADATA_KEY,
            store_key=WORKER_SANDBOX_STORE_METADATA_KEY,
            sandbox_id=sandbox.sandbox_id,
        )
    return sandbox


def ensure_worker_sandbox(
    settings: Settings,
    *,
    template_name: str,
    timeout_seconds: int,
    worker_max_idle_seconds: float | None,
    log_level: str,
    sandbox_id: str | None = None,
) -> Sandbox:
    if sandbox_id:
        try:
            return start_worker_sandbox(
                settings,
                template_name=template_name,
                timeout_seconds=timeout_seconds,
                worker_max_idle_seconds=worker_max_idle_seconds,
                log_level=log_level,
                sandbox_id=sandbox_id,
            )
        except Exception:
            logger.exception(
                "failed to connect worker sandbox %s; creating a replacement",
                sandbox_id,
            )

    return start_worker_sandbox(
        settings,
        template_name=template_name,
        timeout_seconds=timeout_seconds,
        worker_max_idle_seconds=worker_max_idle_seconds,
        log_level=log_level,
    )


def start_webhook_server_process(
    sandbox: Sandbox,
    settings: Settings,
    *,
    worker_max_idle_seconds: float | None,
    log_level: str,
    port: int,
) -> None:
    envs = {
        "ANTHROPIC_ENVIRONMENT_ID": settings.require_anthropic_environment_id(),
        "ANTHROPIC_ENVIRONMENT_KEY": settings.require_anthropic_environment_key(),
        "WORKER_MAX_IDLE_SECONDS": "none"
        if worker_max_idle_seconds is None
        else str(worker_max_idle_seconds),
        "LOG_LEVEL": log_level,
    }
    if settings.anthropic_webhook_signing_key:
        envs["ANTHROPIC_WEBHOOK_SIGNING_KEY"] = settings.anthropic_webhook_signing_key
    uvicorn_command = (
        "python -m uvicorn anthropic_managed_agents_e2b.webhook_runtime:app "
        f"--host 0.0.0.0 --port {port} > {shlex.quote(REMOTE_WEBHOOK_LOG)} "
        "2>&1 < /dev/null &"
    )
    wrapper = f"""
    set -eu
    cd /opt/anthropic-managed-agents
    nohup {uvicorn_command}
    printf '%s\\n' "$!" > {shlex.quote(REMOTE_WEBHOOK_PID)}
    """
    result = sandbox.commands.run(f"bash -lc {shlex.quote(wrapper)}", envs=envs, timeout=15)
    if result.exit_code not in (0, None):
        raise RuntimeError(f"webhook server start failed:\n{result.stdout}\n{result.stderr}")


def write_webhook_config(
    sandbox: Sandbox,
    settings: Settings,
    *,
    worker_max_idle_seconds: float | None,
    log_level: str,
) -> None:
    sandbox.files.write(
        REMOTE_ENVIRONMENT_ID,
        f"{settings.require_anthropic_environment_id()}\n",
    )
    sandbox.files.write(
        REMOTE_ENVIRONMENT_KEY,
        f"{settings.require_anthropic_environment_key()}\n",
    )
    sandbox.files.write(
        REMOTE_WORKER_MAX_IDLE_SECONDS,
        "none\n" if worker_max_idle_seconds is None else f"{worker_max_idle_seconds}\n",
    )
    sandbox.files.write(REMOTE_LOG_LEVEL, f"{log_level}\n")
    if settings.anthropic_webhook_signing_key:
        sandbox.files.write(
            REMOTE_WEBHOOK_SIGNING_KEY,
            f"{settings.anthropic_webhook_signing_key}\n",
        )


def start_webhook_server_sandbox(
    settings: Settings,
    *,
    template_name: str,
    timeout_seconds: int,
    worker_max_idle_seconds: float | None,
    log_level: str,
    port: int,
    sandbox_id: str | None = None,
) -> Sandbox:
    settings.require_anthropic_environment_id()
    settings.require_anthropic_environment_key()

    if sandbox_id:
        sandbox = Sandbox.connect(sandbox_id, timeout=timeout_seconds)
    else:
        sandbox = Sandbox.create(
            template_name,
            timeout=timeout_seconds,
            lifecycle={"on_timeout": "pause", "auto_resume": True},
            metadata={
                "managed_by": "anthropic-managed-agents-e2b-webhook",
                "anthropic.environment_id": settings.anthropic_environment_id or "",
            },
        )

    upload_worker(sandbox)
    write_webhook_config(
        sandbox,
        settings,
        worker_max_idle_seconds=worker_max_idle_seconds,
        log_level=log_level,
    )
    start_webhook_server_process(
        sandbox,
        settings,
        worker_max_idle_seconds=worker_max_idle_seconds,
        log_level=log_level,
        port=port,
    )
    if settings.anthropic_api_key:
        add_sandbox_to_metadata_store(
            api_key=settings.anthropic_api_key,
            environment_id=settings.require_anthropic_environment_id(),
            legacy_key=WEBHOOK_SANDBOX_METADATA_KEY,
            store_key=WEBHOOK_SANDBOX_STORE_METADATA_KEY,
            sandbox_id=sandbox.sandbox_id,
        )
    return sandbox


def stop_worker_sandbox(settings: Settings, sandbox_id: str) -> None:
    Sandbox.kill(sandbox_id)
    JsonSandboxStore().remove_sandbox(sandbox_id=sandbox_id)
    if settings.anthropic_api_key and settings.anthropic_environment_id:
        clear_matching_sandbox_metadata(
            api_key=settings.anthropic_api_key,
            environment_id=settings.anthropic_environment_id,
            sandbox_id=sandbox_id,
        )


def upload_file_to_sandbox(*, sandbox_id: str, local_path: Path, remote_path: str) -> str:
    sandbox = Sandbox.connect(sandbox_id)
    sandbox.files.write(remote_path, local_path.read_bytes())
    return remote_path

File: examples/anthropic-managed-agents/javascript/src/sandbox-worker.ts
import { readFile } from "node:fs/promises";
import { resolve } from "node:path";
import { Sandbox } from "e2b";

import {
  DEFAULT_LOG_LEVEL,
  DEFAULT_SANDBOX_TIMEOUT_SECONDS,
  DEFAULT_TEMPLATE_NAME,
  DEFAULT_WEBHOOK_PORT,
  DEFAULT_WORKER_MAX_IDLE_SECONDS,
  REMOTE_ANTHROPIC_API_KEY,
  REMOTE_APP_SANDBOX_ROUTING_SCOPE,
  REMOTE_APP_WEBHOOK_ADMIN_TOKEN,
  REMOTE_CONFIG_DIR,
  REMOTE_E2B_API_KEY,
  REMOTE_ENVIRONMENT_ID,
  REMOTE_ENVIRONMENT_KEY,
  REMOTE_LOG_LEVEL,
  REMOTE_LOG,
  REMOTE_PID,
  REMOTE_SRC_DIR,
  REMOTE_TSX,
  REMOTE_WEBHOOK,
  REMOTE_WEBHOOK_LOG,
  REMOTE_WEBHOOK_PID,
  REMOTE_WEBHOOK_SIGNING_KEY,
  REMOTE_WORKER_MAX_IDLE_SECONDS,
  REMOTE_WORKDIR,
  REMOTE_WORKER,
} from "./constants.js";
import {
  WEBHOOK_SANDBOX_METADATA_KEY,
  WEBHOOK_SANDBOX_STORE_METADATA_KEY,
  WORKER_SANDBOX_METADATA_KEY,
  WORKER_SANDBOX_STORE_METADATA_KEY,
  addSandboxToMetadataStore,
  clearMatchingSandboxMetadata,
} from "./environment.js";
import { JsonSandboxStore } from "./app-sandbox-store.js";
import { exampleRoot, requireSetting, type Settings } from "./settings.js";

export type WorkerOptions = {
  templateName?: string;
  timeoutSeconds?: number;
  workerMaxIdleSeconds?: number | null;
  logLevel?: string;
  sandboxId?: string;
  sandboxIds?: string[];
  workId?: string;
  sessionId?: string;
};

export type WebhookOptions = WorkerOptions & {
  port?: number;
};

async function createOrConnectWorkerSandbox(settings: Settings, options: WorkerOptions) {
  const timeoutMs = (options.timeoutSeconds ?? DEFAULT_SANDBOX_TIMEOUT_SECONDS) * 1000;
  if (options.sandboxId) {
    return Sandbox.connect(options.sandboxId, { timeoutMs });
  }

  return Sandbox.create(options.templateName ?? DEFAULT_TEMPLATE_NAME, {
    timeoutMs,
    lifecycle: { onTimeout: "pause", autoResume: true },
    metadata: {
      managed_by: "anthropic-managed-agents-e2b-js",
      "anthropic.environment_id": settings.anthropicEnvironmentId ?? "",
    },
  });
}

async function uploadRuntime(sandbox: Sandbox) {
  await sandbox.commands.run(`mkdir -p ${REMOTE_SRC_DIR}`, { timeoutMs: 15_000 });
  await sandbox.files.write([
    {
      path: REMOTE_WORKER,
      data: await readFile(resolve(exampleRoot, "src", "worker-runtime.ts"), "utf8"),
    },
    {
      path: REMOTE_WEBHOOK,
      data: await readFile(resolve(exampleRoot, "src", "webhook-runtime.ts"), "utf8"),
    },
  ]);
}

function workerEnv(settings: Settings, options: WorkerOptions) {
  return {
    ANTHROPIC_ENVIRONMENT_ID: requireSetting(
      settings.anthropicEnvironmentId,
      "ANTHROPIC_ENVIRONMENT_ID",
    ),
    ANTHROPIC_ENVIRONMENT_KEY: requireSetting(
      settings.anthropicEnvironmentKey,
      "ANTHROPIC_ENVIRONMENT_KEY",
    ),
    WORKER_MAX_IDLE_SECONDS:
      options.workerMaxIdleSeconds === null
        ? "none"
        : String(options.workerMaxIdleSeconds ?? DEFAULT_WORKER_MAX_IDLE_SECONDS),
    LOG_LEVEL: (options.logLevel ?? DEFAULT_LOG_LEVEL).toUpperCase(),
    ...(options.workId ? { ANTHROPIC_WORK_ID: options.workId } : {}),
    ...(options.sessionId ? { ANTHROPIC_SESSION_ID: options.sessionId } : {}),
  };
}

export async function workerProcessIsRunning(sandbox: Sandbox) {
  const check =
    `test -f ${REMOTE_PID} && ` +
    `pid="$(cat ${REMOTE_PID})" && ` +
    `test -n "$pid" && ` +
    `kill -0 "$pid"`;
  try {
    const result = await sandbox.commands.run(`bash -lc ${JSON.stringify(check)}`, {
      timeoutMs: 5_000,
    });
    return result.exitCode === 0;
  } catch {
    return false;
  }
}

export async function ensureWorkerProcess(
  sandbox: Sandbox,
  settings: Settings,
  options: WorkerOptions = {},
) {
  await uploadRuntime(sandbox);
  const handlesClaimedWork = Boolean(options.workId || options.sessionId);
  if (!handlesClaimedWork && await workerProcessIsRunning(sandbox)) {
    return;
  }
  const handle = await sandbox.commands.run(`bash -lc ${JSON.stringify(`exec ${REMOTE_TSX} ${REMOTE_WORKER} >> ${REMOTE_LOG} 2>&1`)}`, {
    background: true,
    cwd: REMOTE_WORKDIR,
    envs: workerEnv(settings, options),
  });
  await sandbox.files.write(REMOTE_PID, `${handle.pid}\n`);
  await handle.disconnect();
}

export async function startWorkerSandbox(settings: Settings, options: WorkerOptions = {}) {
  requireSetting(settings.anthropicEnvironmentId, "ANTHROPIC_ENVIRONMENT_ID");
  requireSetting(settings.anthropicEnvironmentKey, "ANTHROPIC_ENVIRONMENT_KEY");

  const sandbox = await createOrConnectWorkerSandbox(settings, options);
  await ensureWorkerProcess(sandbox, settings, options);
  if (settings.anthropicApiKey) {
    await addSandboxToMetadataStore({
      apiKey: settings.anthropicApiKey,
      environmentId: requireSetting(settings.anthropicEnvironmentId, "ANTHROPIC_ENVIRONMENT_ID"),
      legacyKey: WORKER_SANDBOX_METADATA_KEY,
      storeKey: WORKER_SANDBOX_STORE_METADATA_KEY,
      sandboxId: sandbox.sandboxId,
    });
  }

  return sandbox;
}

export async function ensureWorkerSandbox(settings: Settings, options: WorkerOptions = {}) {
  const candidateIds = [
    ...new Set([...(options.sandboxIds ?? []), options.sandboxId].filter((item): item is string => Boolean(item))),
  ];
  for (const candidateId of candidateIds) {
    try {
      return await startWorkerSandbox(settings, { ...options, sandboxId: candidateId });
    } catch (error) {
      console.warn(`failed to connect worker sandbox ${candidateId}; trying next candidate`, error);
    }
  }

  try {
    return await startWorkerSandbox(settings, options);
  } catch (error) {
    if (!options.sandboxId) {
      throw error;
    }
    console.warn(`failed to connect worker sandbox ${options.sandboxId}; creating a replacement`, error);
    return startWorkerSandbox(settings, { ...options, sandboxId: undefined });
  }
}

export async function startWebhookServerSandbox(settings: Settings, options: WebhookOptions = {}) {
  requireSetting(settings.anthropicEnvironmentId, "ANTHROPIC_ENVIRONMENT_ID");
  requireSetting(settings.anthropicEnvironmentKey, "ANTHROPIC_ENVIRONMENT_KEY");

  const timeoutMs = (options.timeoutSeconds ?? DEFAULT_SANDBOX_TIMEOUT_SECONDS) * 1000;
  const sandbox = options.sandboxId
    ? await Sandbox.connect(options.sandboxId, { timeoutMs })
    : await Sandbox.create(options.templateName ?? DEFAULT_TEMPLATE_NAME, {
        timeoutMs,
        envs: {
          ...workerEnv(settings, options),
          WEBHOOK_PORT: String(options.port ?? DEFAULT_WEBHOOK_PORT),
          ...(settings.anthropicWebhookSigningKey
            ? { ANTHROPIC_WEBHOOK_SIGNING_KEY: settings.anthropicWebhookSigningKey }
            : {}),
        },
        lifecycle: { onTimeout: "pause", autoResume: true },
        metadata: {
          managed_by: "anthropic-managed-agents-e2b-js-webhook",
          "anthropic.environment_id": settings.anthropicEnvironmentId ?? "",
        },
      });

  await uploadRuntime(sandbox);
  const envs: Record<string, string> = {
    ...workerEnv(settings, options),
    WEBHOOK_PORT: String(options.port ?? DEFAULT_WEBHOOK_PORT),
  };
  if (settings.anthropicWebhookSigningKey) {
    envs.ANTHROPIC_WEBHOOK_SIGNING_KEY = settings.anthropicWebhookSigningKey;
  }
  await sandbox.commands.run(`mkdir -p ${REMOTE_CONFIG_DIR} && chmod 700 ${REMOTE_CONFIG_DIR}`, {
    timeoutMs: 5_000,
  });
  await sandbox.files.write([
    ...(settings.e2bApiKey ? [{ path: REMOTE_E2B_API_KEY, data: `${settings.e2bApiKey}\n` }] : []),
    ...(settings.anthropicApiKey
      ? [{ path: REMOTE_ANTHROPIC_API_KEY, data: `${settings.anthropicApiKey}\n` }]
      : []),
    { path: REMOTE_ENVIRONMENT_ID, data: `${envs.ANTHROPIC_ENVIRONMENT_ID}\n` },
    { path: REMOTE_ENVIRONMENT_KEY, data: `${envs.ANTHROPIC_ENVIRONMENT_KEY}\n` },
    {
      path: REMOTE_WORKER_MAX_IDLE_SECONDS,
      data: `${envs.WORKER_MAX_IDLE_SECONDS ?? DEFAULT_WORKER_MAX_IDLE_SECONDS}\n`,
    },
    { path: REMOTE_LOG_LEVEL, data: `${envs.LOG_LEVEL ?? DEFAULT_LOG_LEVEL}\n` },
    ...(settings.anthropicWebhookSigningKey
      ? [{ path: REMOTE_WEBHOOK_SIGNING_KEY, data: `${settings.anthropicWebhookSigningKey}\n` }]
      : []),
    ...(settings.appWebhookAdminToken
      ? [{ path: REMOTE_APP_WEBHOOK_ADMIN_TOKEN, data: `${settings.appWebhookAdminToken}\n` }]
      : []),
    ...(settings.appSandboxRoutingScope
      ? [{ path: REMOTE_APP_SANDBOX_ROUTING_SCOPE, data: `${settings.appSandboxRoutingScope}\n` }]
      : []),
  ]);
  await sandbox.commands.run(`chmod 600 ${REMOTE_CONFIG_DIR}/*`, { timeoutMs: 5_000 });

  const healthUrl = `http://127.0.0.1:${options.port ?? DEFAULT_WEBHOOK_PORT}/health`;
  let webhookServerReady = false;
  try {
    await sandbox.commands.run(`curl --fail --silent --show-error ${healthUrl}`, {
      timeoutMs: 5_000,
    });
    webhookServerReady = true;
  } catch {
    // Older/local templates do not have a start command; start the server below.
  }

  if (!webhookServerReady) {
    const handle = await sandbox.commands.run(
      `bash -lc ${JSON.stringify(`exec ${REMOTE_TSX} ${REMOTE_WEBHOOK} >> ${REMOTE_WEBHOOK_LOG} 2>&1`)}`,
      {
        background: true,
        cwd: REMOTE_WORKDIR,
        envs,
      },
    );
    await sandbox.files.write(REMOTE_WEBHOOK_PID, `${handle.pid}\n`);
    await handle.disconnect();
  }

  if (settings.anthropicApiKey) {
    await addSandboxToMetadataStore({
      apiKey: settings.anthropicApiKey,
      environmentId: requireSetting(settings.anthropicEnvironmentId, "ANTHROPIC_ENVIRONMENT_ID"),
      legacyKey: WEBHOOK_SANDBOX_METADATA_KEY,
      storeKey: WEBHOOK_SANDBOX_STORE_METADATA_KEY,
      sandboxId: sandbox.sandboxId,
    });
  }

  if (webhookServerReady) {
    return sandbox;
  }

  const attempts = Array.from({ length: 30 }, (_, index) => index + 1).join(" ");
  const healthCheck =
    `for i in ${attempts}; do ` +
    `curl --fail --silent --show-error ${healthUrl} && exit 0; ` +
    `sleep 1; ` +
    `done; ` +
    `echo "webhook log:" >&2; ` +
    `tail -100 ${REMOTE_WEBHOOK_LOG} >&2 || true; ` +
    `exit 1`;
  try {
    await sandbox.commands.run(`bash -lc ${JSON.stringify(healthCheck)}`, {
      timeoutMs: 35_000,
    });
  } catch (error) {
    const result = error as { stdout?: string; stderr?: string };
    const output = [result.stdout, result.stderr].filter(Boolean).join("\n");
    const message = output || (error instanceof Error ? error.message : String(error));
    throw new Error(`webhook server health check failed: ${healthUrl}\n${message}`);
  }

  return sandbox;
}

export async function stopWorkerSandbox(settings: Settings, sandboxId: string) {
  const sandbox = await Sandbox.connect(sandboxId);
  await sandbox.kill();
  await new JsonSandboxStore().removeSandbox(sandboxId);
  if (settings.anthropicApiKey && settings.anthropicEnvironmentId) {
    await clearMatchingSandboxMetadata({
      apiKey: settings.anthropicApiKey,
      environmentId: settings.anthropicEnvironmentId,
      sandboxId,
    });
  }
}

export async function uploadFileToSandbox({
  sandboxId,
  localPath,
  remotePath,
}: {
  sandboxId: string;
  localPath: string;
  remotePath: string;
}) {
  const sandbox = await Sandbox.connect(sandboxId);
  const data = await readFile(localPath);
  const arrayBuffer = data.buffer.slice(data.byteOffset, data.byteOffset + data.byteLength);
  await sandbox.files.write(remotePath, arrayBuffer);
  return remotePath;
}
